src/main/java/io/dgraph/DgraphClient.java

```diff
@@ -31,6 +41,154 @@ public class DgraphClient {
+    }
+
+    if (username != null && (password == null || password.isEmpty())) {
+      throw new IllegalArgumentException(
+          "Invalid connection string: password required when username is provided");
+    }
+
+    // Parse parameters into a Map (crazy that there's no built-in support for this in net.URL)
+    Map<String, String> params = new HashMap<>();
+    if (url.getQuery() != null) {
+      String[] pairs = url.getQuery().split("&");
+      for (String pair : pairs) {
+        int idx = pair.indexOf("=");
+        if (idx > 0) {
+          try {
+            String key = URLDecoder.decode(pair.substring(0, idx), StandardCharsets.UTF_8.toString());
+            String value = URLDecoder.decode(pair.substring(idx + 1), StandardCharsets.UTF_8.toString());
+            params.put(key, value);
+          } catch (UnsupportedEncodingException e) {
+            throw new AssertionError(e);
+          }
+        }
+      }
+    }
+
+    ManagedChannelBuilder<?> channelBuilder = ManagedChannelBuilder
+        .forAddress(host, port);
+
+    if (params.containsKey("sslmode")) {
+      String sslmode = params.get("sslmode");
+      if ("disable".equals(sslmode)) {
+        channelBuilder.usePlaintext();
+      } else if ("require".equals(sslmode)) {
+        // create a new channel builder for tls minus the CA checks
+        try {
+          SslContext sslContext = GrpcSslContexts.forClient()
+            .trustManager(InsecureTrustManagerFactory.INSTANCE)
+            .build();
+          channelBuilder = NettyChannelBuilder.forAddress(host, port)
+            .sslContext(sslContext);
+        } catch (SSLException e) {
+          throw e;
+        }
+      } else if ("verify-ca".equals(sslmode)) {
+        channelBuilder.useTransportSecurity();
+      } else {
+        throw new IllegalArgumentException("Invalid sslmode: " + sslmode);
+      }
+    } else {
+      channelBuilder.usePlaintext();
+    }
+
+    if (params.containsKey("apikey") && params.containsKey("bearertoken")) {
+      throw new IllegalArgumentException(
+          "apikey and bearertoken cannot both be provided");
+    }
+
+    String authHeader = null;
+    if (params.containsKey("apikey")) {
+      authHeader = params.get("apikey");
+    } else if (params.containsKey("bearertoken")) {
+      authHeader = "Bearer " + params.get("bearertoken");
+    }
+
+    DgraphGrpc.DgraphStub stub;
+    if (authHeader != null) {
+      Metadata metadata = new Metadata();
+      metadata.put(
+          Metadata.Key.of(gRPC_AUTHORIZATION_HEADER_NAME, Metadata.ASCII_STRING_MARSHALLER),
+          authHeader);
+      stub = DgraphGrpc.newStub(channelBuilder.build())
+          .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata));
+    } else {
+      stub = DgraphGrpc.newStub(channelBuilder.build());
+    }
+
+    DgraphClient client = new DgraphClient(stub);
+
+    if (username != null) {
+      client.login(username, password);
+    }
+    return client;
+  }
```

Comment:
This is a long function, we can break this down into withOption set of private functions that we can also use later for NewClient function